levels/lovelace_tower/events.js

```diff
@@ -1,353 +1,162 @@
-
-// If No: 
-// [] Exit door from library is locked with Operator observation
-
-// If Yes:
-// [] Sparkle animation appears on exit door and is interactable
-// [] Previous entry door back into Tower becomes locked
-
-
-// UNLOCKING
-
-  const unlockObject = (group) => {
-    world.showEntities(
-      ({ instance }) => instance.group === group || instance.key == group
-    );
-    if (!worldState.unlockedEntities.includes(group))
-      worldState.unlockedEntities.push(group);
-  };
-
-  const unlockTransition = (group) => {
-    world.enableTransitionAreas(({ instance }) => instance.name === group);
-    if (!worldState.unlockedTransitions.includes(group))
-      worldState.unlockedTransitions.push(group);
-  };
-
-  const destroyObject = (group) => {
-    world.destroyEntities(
-      ({ instance }) => instance.group === group || instance.key == group
-    );
-    if (!worldState.destroyedEntities.includes(group))
-      worldState.destroyedEntities.push(group);
-  };
-
-  const unhackObject = (group) => {
-    world.forEachEntities(
-      ({ instance }) => instance.group === group || instance.key == group,
-      (entity) => {
-        entity.hackable = false;
-      }
-    );
-    if (!worldState.unhackableEntities.includes(group))
-      worldState.unhackableEntities.push(group);
-  };
-
-  const openDoor = (group) => {
-    world.forEachEntities(group, (door) => {
-      door.state.fsm.action("open");
-    });
-  };
-
-
-// NPC CONVERSATIONS
-
-  const runObjectNotification = ({ target: { key = "default" } }) => {
-    world.startConversation(key, key + ".png");
-  };
-
-  const runNpcChecks = (event) => {
-    npcChecks[event.target.key]();
-  };
-
-// MISSION OBJECTIVES
-
-// Add specific objectives for Lovelace Tower and Secret Lovelace Library
-  const objectives = {
-    lovelace_key: () => obtainKeySpellComplete("lovelace_spell"),
-    hopper_key: () => obtainKeySpellComplete("hopper_spell"),
-    neumann_key: () => obtainKeySpellComplete("neumann_spell"),
-    turing_key: () => obtainKeySpellComplete("turing_spell"),
-  };
-
-// When the player has retrieved the House Lovelace spell, they are able to enter the Secret Library
-// They are also able to open the library door that leads back to the Main Hall
-// Eventually they will also be able to use the spell to gain access to the next house from the Main Hall
-  const obtainKeySpellComplete = (object) => {
-    unlockObject(object);
-    if (!worldState.insideCatacombs.keySpellsObtained.includes(object))
-      worldState.insideCatacombs.keySpellsObtained.push(object);
-
-    if (worldState.insideCatacombs.keySpellsObtained.length == 4) {
-      world.grantItems(["magic_key"]);
-      unlockObject("magic_key");
+  const worldState = merge(
+    INITIAL_STATE,
+    world.getState(LOVELACE_TOWER_STATE_KEY)
+  );
+
+  // Top object pieces need to be longer so they always depth sort
+  // above the player, even when the player would normally be depth
+  // sorted above them.
```

Comment:
They need to be above the player even when they'd be above the player?

Reply:
I think we might be able to use the newer [zIndex](https://github.com/twilio/twilioquest/blob/3.2_next/src/js/app/components/game/DynamicLevel.js#L294-L298) property for this usecase now.

I think this is used in TOIK exterior level on layers. I believe it can be applied to objects directly as well though.

Reply:
> They need to be above the player even when they'd be above the player?

This is to say, render top objects above the player even when they normally wouldn't. Bit of a lazily edited version of [this](https://github.com/TwilioQuest/twilioquest-dev-fundamentals/blob/main/levels/tower_of_knowledge/events.js#:~:text=//%20Top%20shelf%20pieces,sorted%20above%20them.) haha


> I think we might be able to use the newer [zIndex](https://github.com/twilio/twilioquest/blob/3.2_next/src/js/app/components/game/DynamicLevel.js#L294-L298) property for this usecase now.
> 
> I think this is used in TOIK exterior level on layers. I believe it can be applied to objects directly as well though.

Just tried it and it works! I'll go ahead and use that approach instead